File: electron/src/core/replay.ts
/**
 * Replay system for historical session playback.
 */

import * as fs from 'fs';
import * as path from 'path';
import log from 'electron-log';
import { TrackerRecord } from './models';
import { CSVParser } from './parser';
import { StateManager } from './state';
import { getTestSessions } from './library-store';

export interface ReplaySession {
  session_id: string;
  name: string;
  path: string;
  duration_seconds: number;
  start_time: string;
  end_time: string;
  tracker_ids: string[];
  file_count: number;
  total_records: number;
  size_bytes: number;
}

export interface FrameGroup {
  frame_index: number;
  timestamp: string;
  records: Map<string, TrackerRecord>;
  duration_seconds: number;
}

export type BroadcastCallback = (msg: Record<string, unknown>) => void;

export class SessionLoader {
  private parser = new CSVParser();

  constructor(private logRoot: string) {}

  async scanSessions(): Promise<ReplaySession[]> {
    const sessions: ReplaySession[] = [];

    if (!fs.existsSync(this.logRoot)) return sessions;

    try {
      // Check if logRoot itself contains CSV files (treat as a session if so)
      const csvInRoot = fs.readdirSync(this.logRoot)
        .filter((f) => path.extname(f).toLowerCase() === '.csv');

      if (csvInRoot.length > 0) {
        const session = await this.analyzeSession(this.logRoot);
        if (session) {
          sessions.push(session);
          // NOTE: Don't return early - continue scanning subdirectories for more sessions
        }
      }

      // Scan subdirectories (one level)
      const entries = fs.readdirSync(this.logRoot, { withFileTypes: true });
      for (const entry of entries) {
        if (!entry.isDirectory() || entry.name.startsWith('.')) continue;

        const entryPath = path.join(this.logRoot, entry.name);

        // Check if this directory contains CSV files directly
        const session = await this.analyzeSession(entryPath);
        if (session) {
          sessions.push(session);
        } else {
          // If no CSV files found, scan subdirectories (for test-sessions/ folder structure)
          // This handles: log_root/test-sessions/{session_name}/*.csv
          try {
            const subEntries = fs.readdirSync(entryPath, { withFileTypes: true });
            for (const subEntry of subEntries) {
              if (!subEntry.isDirectory() || subEntry.name.startsWith('.')) continue;
              const subSession = await this.analyzeSession(path.join(entryPath, subEntry.name));
              if (subSession) sessions.push(subSession);
            }
          } catch {
            // Ignore errors scanning subdirectories
          }
        }
      }

      // Also check library-store for completed sessions with live_data_path
      // This bridges the gap between JSON metadata and file-based storage
      try {
        const testSessions = getTestSessions().filter(
          (s) => (s.status === 'completed' || s.status === 'capturing') && s.live_data_path
        );

        for (const ts of testSessions) {
          // Check if already found via file scan
          const alreadyFound = sessions.find((s) => s.path === ts.live_data_path);
          if (!alreadyFound && ts.live_data_path && fs.existsSync(ts.live_data_path)) {
            const session = await this.analyzeSession(ts.live_data_path);
            if (session) {
              // Use the session name from the metadata
              session.name = ts.name;
              session.session_id = ts.id;
              sessions.push(session);
              log.info(`[Replay] Found session from library-store: ${ts.name} at ${ts.live_data_path}`);
            }
          }
        }
      } catch (libError) {
        log.warn('[Replay] Could not check library-store for sessions:', libError);
      }

      // Sort by most recent first
      sessions.sort((a, b) => b.start_time.localeCompare(a.start_time));
      return sessions;
    } catch (e) {
      log.error('Error scanning sessions:', e);
      return [];
    }
  }

  private async analyzeSession(sessionPath: string): Promise<ReplaySession | null> {
    try {
      const csvFiles = fs.readdirSync(sessionPath)
        .filter((f) => path.extname(f).toLowerCase() === '.csv')
        .map((f) => path.join(sessionPath, f));

      if (csvFiles.length === 0) return null;

      let totalSize = 0;
      const trackerIds = new Set<string>();
      const allTimes: number[] = [];
      let totalRecords = 0;

      for (const csvFile of csvFiles) {
        try {
          const stat = fs.statSync(csvFile);
          totalSize += stat.size;

          const content = fs.readFileSync(csvFile, 'utf-8');
          const lines = content.split('\n').filter((l) => l.trim());
          if (lines.length <= 1) continue;

          const header = lines[0];

          // Parse first record
          if (lines.length > 1) {
            const records = this.parser.parseCSVContent(header + '\n' + lines[1]);
            if (records.length > 0) {
              trackerIds.add(records[0].tracker_id);
              const t = new Date(records[0].time_local_received).getTime();
              if (!isNaN(t)) allTimes.push(t);
            }
          }

          // Parse last record
          if (lines.length > 2) {
            const records = this.parser.parseCSVContent(header + '\n' + lines[lines.length - 1]);
            if (records.length > 0) {
              const t = new Date(records[0].time_local_received).getTime();
              if (!isNaN(t)) allTimes.push(t);
            }
          }

          totalRecords += lines.length - 1;
        } catch {
          continue;
        }
      }

      if (allTimes.length === 0) return null;

      const startTime = Math.min(...allTimes);
      const endTime = Math.max(...allTimes);
      const duration = (endTime - startTime) / 1000;

      return {
        session_id: path.basename(sessionPath),
        name: path.basename(sessionPath),
        path: sessionPath,
        duration_seconds: duration,
        start_time: new Date(startTime).toISOString(),
        end_time: new Date(endTime).toISOString(),
        tracker_ids: Array.from(trackerIds).sort(),
        file_count: csvFiles.length,
        total_records: totalRecords,
        size_bytes: totalSize,
      };
    } catch (e) {
      log.error(`Error analyzing session ${sessionPath}:`, e);
      return null;
    }
  }

  async loadTimeline(
    session: ReplaySession,
    selectedTrackers?: string[],
    progressCallback?: (frameIndex: number, totalFrames: number) => void
  ): Promise<FrameGroup[]> {
    const allRecords: TrackerRecord[] = [];

    const csvFiles = fs.readdirSync(session.path)
      .filter((f) => path.extname(f).toLowerCase() === '.csv')
      .map((f) => path.join(session.path, f));

    for (const csvFile of csvFiles) {
      try {
        const content = fs.readFileSync(csvFile, 'utf-8');
        const records = this.parser.parseCSVContent(content);

        for (const record of records) {
          if (!selectedTrackers || selectedTrackers.includes(record.tracker_id)) {
            allRecords.push(record);
          }
        }
      } catch (e) {
        log.error(`Error loading ${csvFile}:`, e);
      }
    }

    if (allRecords.length === 0) return [];

    // Sort by time
    allRecords.sort((a, b) => a.time_local_received.localeCompare(b.time_local_received));

    // Use optimized async buildFrames with progress reporting
    return await this.buildFrames(allRecords, 0.5, progressCallback);
  }

  /**
   * Build frame groups from records - OPTIMIZED with index tracking
   * Uses O(n) algorithm instead of O(n³) by tracking position per tracker.
   * Includes progress reporting and chunked processing.
   */
  private async buildFrames(
    records: TrackerRecord[],
    frameInterval = 0.5,
    progressCallback?: (frameIndex: number, totalFrames: number) => void
  ): Promise<FrameGroup[]> {
    if (records.length === 0) return [];

    const frames: FrameGroup[] = [];
    const startTime = new Date(records[0].time_local_received).getTime();
    const endTime = new Date(records[records.length - 1].time_local_received).getTime();

    // Group by tracker and pre-compute timestamps for efficiency
    const byTracker = new Map<string, { record: TrackerRecord; timestamp: number }[]>();
    for (const record of records) {
      const list = byTracker.get(record.tracker_id) || [];
      list.push({
        record,
        timestamp: new Date(record.time_local_received).getTime(),
      });
      byTracker.set(record.tracker_id, list);
    }

    // Sort each tracker's records by timestamp (they should already be sorted, but ensure it)
    for (const [, trackerRecords] of byTracker) {
      trackerRecords.sort((a, b) => a.timestamp - b.timestamp);
    }

    const intervalMs = frameInterval * 1000;
    let currentTime = startTime;
    let frameIndex = 0;
    const CHUNK_SIZE = 100; // Yield more frequently for better responsiveness
    const PROGRESS_INTERVAL = 50; // Report progress every 50 frames

    // Calculate total frames for progress reporting
    const totalFrames = Math.ceil((endTime - startTime) / intervalMs);

    // Track current index per tracker for O(1) lookup instead of O(n) linear search
    const trackerIndices = new Map<string, number>();
    for (const trackerId of byTracker.keys()) {
      trackerIndices.set(trackerId, 0);
    }

    log.info(`[Replay] Building ${totalFrames} frames for ${byTracker.size} trackers (${records.length} records)`);

    while (currentTime <= endTime) {
      const frameRecords = new Map<string, TrackerRecord>();

      for (const [trackerId, trackerRecords] of byTracker) {
        // Get current index for this tracker (optimized: O(1) instead of O(n))
        let idx = trackerIndices.get(trackerId) || 0;

        // Move forward through records until we find one past current time
        // This is O(1) amortized because we only move forward
        while (
          idx < trackerRecords.length - 1 &&
          trackerRecords[idx + 1].timestamp <= currentTime
        ) {
          idx++;
        }

        // Update the tracked index for next frame
        trackerIndices.set(trackerId, idx);

        // Check if we have a valid record within tolerance
        const currentRecord = trackerRecords[idx];
        if (currentRecord) {
          const delta = Math.abs(currentRecord.timestamp - currentTime);

          // Also check the next record if available for a potentially closer match
          if (idx < trackerRecords.length - 1) {
            const nextRecord = trackerRecords[idx + 1];
            const nextDelta = Math.abs(nextRecord.timestamp - currentTime);
            if (nextDelta < delta && nextDelta < intervalMs * 2) {
              frameRecords.set(trackerId, nextRecord.record);
            } else if (delta < intervalMs * 2) {
              frameRecords.set(trackerId, currentRecord.record);
            }
          } else if (delta < intervalMs * 2) {
            frameRecords.set(trackerId, currentRecord.record);
          }
        }
      }

      if (frameRecords.size > 0) {
        frames.push({
          frame_index: frameIndex,
          timestamp: new Date(currentTime).toISOString(),
          records: frameRecords,
          duration_seconds: frameInterval,
        });
      }

      currentTime += intervalMs;
      frameIndex++;

      // Report progress periodically
      if (progressCallback && frameIndex % PROGRESS_INTERVAL === 0) {
        progressCallback(frameIndex, totalFrames);
      }

      // Yield to event loop periodically to prevent UI freeze
      if (frameIndex % CHUNK_SIZE === 0) {
        await new Promise(resolve => setImmediate(resolve));
      }
    }

    // Final progress report
    if (progressCallback) {
      progressCallback(totalFrames, totalFrames);
    }

    log.info(`[Replay] Built ${frames.length} frames for ${byTracker.size} trackers`);
    return frames;
  }
}

export class ReplayEngine {
  currentFrameIndex = 0;
  playbackSpeed = 1.0;
  isPlaying = false;
  private playbackTimer: ReturnType<typeof setTimeout> | null = null;
  private broadcastCallback: BroadcastCallback | null = null;

  constructor(
    private frames: FrameGroup[],
    private stateManager: StateManager
  ) {}

  setBroadcastCallback(callback: BroadcastCallback): void {
    this.broadcastCallback = callback;
  }

  play(): void {
    if (this.isPlaying) return;
    this.isPlaying = true;
    this.scheduleNextFrame();
  }

  pause(): void {
    this.isPlaying = false;
    if (this.playbackTimer) {
      clearTimeout(this.playbackTimer);
      this.playbackTimer = null;
    }
  }

  seek(frameIndex: number): void {
    if (frameIndex < 0 || frameIndex >= this.frames.length) return;

    const wasPlaying = this.isPlaying;
    if (wasPlaying) this.pause();

    this.currentFrameIndex = frameIndex;
    this.emitFrame(this.frames[frameIndex]);

    if (wasPlaying) this.play();
  }

  setSpeed(speed: number): void {
    this.playbackSpeed = Math.max(0.1, Math.min(10.0, speed));
  }

  getCurrentState(): Record<string, unknown> {
    const currentFrame = this.currentFrameIndex < this.frames.length
      ? this.frames[this.currentFrameIndex]
      : null;

    return {
      is_playing: this.isPlaying,
      current_frame: this.currentFrameIndex,
      total_frames: this.frames.length,
      playback_speed: this.playbackSpeed,
      current_time: currentFrame?.timestamp || null,
      progress_percent: this.frames.length > 0
        ? (this.currentFrameIndex / this.frames.length) * 100
        : 0,
    };
  }

  private scheduleNextFrame(): void {
    if (!this.isPlaying || this.currentFrameIndex >= this.frames.length) {
      if (this.currentFrameIndex >= this.frames.length) {
        this.isPlaying = false;
        if (this.broadcastCallback) {
          this.broadcastCallback({ type: 'replay_completed' });
        }
      }
      return;
    }

    const frame = this.frames[this.currentFrameIndex];
    this.emitFrame(frame);

    const sleepTime = Math.max(1, Math.min(1000, (frame.duration_seconds / this.playbackSpeed) * 1000));

    this.playbackTimer = setTimeout(() => {
      this.currentFrameIndex++;

      if (this.broadcastCallback) {
        this.broadcastCallback({
          type: 'replay_progress',
          frame: this.currentFrameIndex,
          total: this.frames.length,
        });
      }

      this.scheduleNextFrame();
    }, sleepTime);
  }

  private emitFrame(frame: FrameGroup): void {
    for (const [, record] of frame.records) {
      this.stateManager.updateTracker(record);
    }
  }
}

File: logtail_dashboard/replay.py
"""Replay system for historical session playback."""

import asyncio
import logging
from dataclasses import dataclass
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional

from .models import TrackerRecord
from .parser import CSVParser
from .state import StateManager

logger = logging.getLogger(__name__)


@dataclass
class ReplaySession:
    """Metadata for a historical test session."""

    session_id: str
    name: str
    path: Path
    duration_seconds: float
    start_time: datetime
    end_time: datetime
    tracker_ids: list[str]
    file_count: int
    total_records: int
    size_bytes: int


@dataclass
class FrameGroup:
    """Single synchronized frame in timeline."""

    frame_index: int
    timestamp: datetime
    records: dict[str, TrackerRecord]  # tracker_id -> record
    duration_seconds: float = 0.033  # Default ~30fps


class SessionLoader:
    """Load and parse historical sessions."""

    def __init__(self, log_root: Path):
        self.log_root = Path(log_root)
        self.parser = CSVParser()
        # Cache sessions to avoid re-scanning on every load
        self._session_cache: dict[str, ReplaySession] = {}
        self._cache_timestamp: Optional[datetime] = None
        self._cache_ttl_seconds = 60  # Cache valid for 60 seconds
        # Thread safety lock for cache operations
        self._cache_lock = asyncio.Lock()

    async def scan_sessions(self, force_refresh: bool = False) -> list[ReplaySession]:
        """Discover all available sessions.

        Uses a cache to avoid re-scanning filesystem on every request.
        Cache is invalidated after _cache_ttl_seconds or on force_refresh.
        Thread-safe using asyncio lock.
        """
        async with self._cache_lock:
            # Check if cache is valid
            if not force_refresh and self._cache_timestamp:
                cache_age = (datetime.now() - self._cache_timestamp).total_seconds()
                if cache_age < self._cache_ttl_seconds and self._session_cache:
                    logger.debug(f"Using cached sessions ({len(self._session_cache)} sessions, {cache_age:.1f}s old)")
                    return list(self._session_cache.values())

            sessions = []

            if not self.log_root.exists():
                logger.warning(f"Log root does not exist: {self.log_root}")
                return sessions

            try:
                # First check if log_root itself contains CSV files (direct folder approach)
                csv_in_root = list(self.log_root.glob("*.csv"))
                if csv_in_root:
                    session = await self._analyze_session(self.log_root)
                    if session:
                        sessions.append(session)
                        # Update cache
                        self._session_cache = {session.session_id: session}
                        self._cache_timestamp = datetime.now()
                        return sessions  # Return immediately if direct folder has CSVs

                # Otherwise scan subdirectories (legacy approach)
                for item in self.log_root.iterdir():
                    if not item.is_dir() or item.name.startswith("."):
                        continue

                    session = await self._analyze_session(item)
                    if session:
                        sessions.append(session)

                # Sort by most recent first
                sessions.sort(key=lambda s: s.start_time, reverse=True)

                # Update cache
                self._session_cache = {s.session_id: s for s in sessions}
                self._cache_timestamp = datetime.now()
                logger.info(f"Scanned and cached {len(sessions)} sessions")

                return sessions

            except Exception as e:
                logger.error(f"Error scanning sessions: {e}")
                return []

    def get_cached_session(self, session_id: str) -> Optional[ReplaySession]:
        """Get a session from cache by ID without re-scanning.

        This method is used by load_replay_session to avoid race conditions.
        """
        return self._session_cache.get(session_id)

    def invalidate_cache(self):
        """Invalidate the session cache, forcing a re-scan on next request."""
        self._session_cache = {}
        self._cache_timestamp = None
        logger.info("Session cache invalidated")

    async def _analyze_session(self, session_path: Path) -> Optional[ReplaySession]:
        """Analyze a session directory."""
        try:
            logger.debug(f"Analyzing session: {session_path}")
            csv_files = list(session_path.glob("*.csv"))
            logger.debug(f"Found {len(csv_files)} CSV files in {session_path}")

            if not csv_files:
                return None

            # Quick metadata extraction
            total_size = sum(f.stat().st_size for f in csv_files)
            tracker_ids = set()
            all_times = []
            total_records = 0

            # Parse first and last few records from each file
            for csv_file in csv_files:
                try:
                    # Read first 10 and last 10 lines for timing
                    with open(csv_file, "r") as f:
                        lines = f.readlines()

                    if len(lines) <= 1:  # Skip if only header
                        continue

                    # Parse first and last records using the parser
                    # Create mini CSV content with header + data line
                    header = lines[0]

                    # Parse first record
                    if len(lines) > 1:
                        mini_csv = header + lines[1]
                        records = self.parser.parse_csv_content(mini_csv)
                        if records:
                            tracker_ids.add(records[0].tracker_id)
                            all_times.append(records[0].time_local_received)

                    # Parse last record
                    if len(lines) > 2:
                        mini_csv = header + lines[-1]
                        records = self.parser.parse_csv_content(mini_csv)
                        if records:
                            all_times.append(records[0].time_local_received)

                    total_records += len(lines) - 1  # Exclude header

                except Exception as e:
                    logger.debug(f"Error analyzing {csv_file}: {e}")
                    continue

            if not all_times:
                return None

            start_time = min(all_times)
            end_time = max(all_times)
            duration = (end_time - start_time).total_seconds()

            return ReplaySession(
                session_id=session_path.name,
                name=session_path.name,
                path=session_path,
                duration_seconds=duration,
                start_time=start_time,
                end_time=end_time,
                tracker_ids=sorted(list(tracker_ids)),
                file_count=len(csv_files),
                total_records=total_records,
                size_bytes=total_size,
            )

        except Exception as e:
            logger.error(f"Error analyzing session {session_path}: {e}")
            return None

    async def load_timeline(
        self, session: ReplaySession, selected_trackers: Optional[list[str]] = None
    ) -> list[FrameGroup]:
        """Load and parse all records into synchronized frames."""
        logger.info(f"Loading timeline for session: {session.name}")

        # Collect all records
        all_records = []
        csv_files = list(session.path.glob("*.csv"))

        for csv_file in csv_files:
            try:
                with open(csv_file, "r") as f:
                    content = f.read()

                # Parse the entire CSV file
                records = self.parser.parse_csv_content(content)

                # Filter by selected trackers if specified
                for record in records:
                    if selected_trackers is None or record.tracker_id in selected_trackers:
                        all_records.append(record)

            except Exception as e:
                logger.error(f"Error loading {csv_file}: {e}")
                continue

        if not all_records:
            logger.warning("No records loaded")
            return []

        # Sort by time
        all_records.sort(key=lambda r: r.time_local_received)

        # Build frames
        frames = await self._build_frames(all_records)

        logger.info(f"Loaded {len(frames)} frames with {len(all_records)} total records")

        return frames

    async def _build_frames(
        self, records: list[TrackerRecord], frame_interval: float = 0.5
    ) -> list[FrameGroup]:
        """Organize records into synchronized frames."""
        if not records:
            return []

        frames = []
        start_time = records[0].time_local_received
        end_time = records[-1].time_local_received

        # Group records by tracker
        records_by_tracker = {}
        for record in records:
            if record.tracker_id not in records_by_tracker:
                records_by_tracker[record.tracker_id] = []
            records_by_tracker[record.tracker_id].append(record)

        # Build frames at regular intervals
        current_time = start_time
        frame_index = 0

        while current_time <= end_time:
            frame = FrameGroup(
                frame_index=frame_index, timestamp=current_time, records={}, duration_seconds=frame_interval
            )

            # Find closest record for each tracker
            for tracker_id, tracker_records in records_by_tracker.items():
                # Find record closest to current_time
                best_record = None
                min_delta = None

                for record in tracker_records:
                    delta = abs((record.time_local_received - current_time).total_seconds())

                    if min_delta is None or delta < min_delta:
                        min_delta = delta
                        best_record = record

                # Only include if within reasonable range (2x frame interval)
                if best_record and min_delta < frame_interval * 2:
                    frame.records[tracker_id] = best_record

            if frame.records:  # Only add frames with data
                frames.append(frame)

            current_time += timedelta(seconds=frame_interval)
            frame_index += 1

        return frames


class ReplayEngine:
    """Control playback timeline."""

    def __init__(self, timeline: list[FrameGroup], state_manager: StateManager):
        self.frames = timeline
        self.state_manager = state_manager
        self.current_frame_index = 0
        self.playback_speed = 1.0
        self.is_playing = False
        self._playback_task: Optional[asyncio.Task] = None
        self._broadcast_callback = None

    def set_broadcast_callback(self, callback):
        """Set callback for broadcasting frame updates."""
        self._broadcast_callback = callback

    async def play(self) -> None:
        """Start playback from current frame."""
        if self.is_playing:
            return

        self.is_playing = True
        self._playback_task = asyncio.create_task(self._playback_loop())
        logger.info(f"Playback started at frame {self.current_frame_index}")

    async def pause(self) -> None:
        """Pause playback."""
        self.is_playing = False
        if self._playback_task:
            self._playback_task.cancel()
            try:
                await self._playback_task
            except asyncio.CancelledError:
                pass
        logger.info(f"Playback paused at frame {self.current_frame_index}")

    async def seek(self, frame_index: int) -> None:
        """Jump to specific frame."""
        if frame_index < 0 or frame_index >= len(self.frames):
            logger.warning(f"Invalid frame index: {frame_index}")
            return

        was_playing = self.is_playing

        if was_playing:
            await self.pause()

        self.current_frame_index = frame_index

        # Emit current frame
        await self._emit_frame(self.frames[frame_index])

        logger.info(f"Seeked to frame {frame_index}")

        if was_playing:
            await self.play()

    async def set_speed(self, speed: float) -> None:
        """Change playback speed (0.1x to 10.0x)."""
        # Clamp to valid range
        speed = max(0.1, min(10.0, speed))
        self.playback_speed = speed
        logger.info(f"Playback speed set to {speed}x")

    async def _playback_loop(self) -> None:
        """Main playback loop."""
        try:
            while self.is_playing and self.current_frame_index < len(self.frames):
                frame = self.frames[self.current_frame_index]

                # Emit frame
                await self._emit_frame(frame)

                # Calculate sleep time based on speed
                sleep_time = frame.duration_seconds / self.playback_speed

                # Clamp to prevent blocking
                sleep_time = max(0.001, min(sleep_time, 1.0))

                await asyncio.sleep(sleep_time)

                # Advance
                self.current_frame_index += 1

                # Notify about progress
                if self._broadcast_callback:
                    await self._broadcast_callback(
                        {
                            "type": "replay_progress",
                            "frame": self.current_frame_index,
                            "total": len(self.frames),
                        }
                    )

            # Reached end
            if self.current_frame_index >= len(self.frames):
                self.is_playing = False
                logger.info("Playback completed")

                if self._broadcast_callback:
                    await self._broadcast_callback({"type": "replay_completed"})

        except asyncio.CancelledError:
            logger.info("Playback cancelled")
        except Exception as e:
            logger.error(f"Playback error: {e}")
            self.is_playing = False

    async def _emit_frame(self, frame: FrameGroup) -> None:
        """Emit frame records to state manager."""
        for tracker_id, record in frame.records.items():
            # Update state manager (this triggers WebSocket broadcast)
            self.state_manager.update_tracker(record)

    def get_current_state(self) -> dict:
        """Get current playback state."""
        current_frame = self.frames[self.current_frame_index] if self.current_frame_index < len(self.frames) else None

        return {
            "is_playing": self.is_playing,
            "current_frame": self.current_frame_index,
            "total_frames": len(self.frames),
            "playback_speed": self.playback_speed,
            "current_time": current_frame.timestamp.isoformat() if current_frame else None,
            "progress_percent": (self.current_frame_index / len(self.frames) * 100) if self.frames else 0,
        }
